politeiawww/invoices_test.go

```diff
@@ -0,0 +1,281 @@
+// Copyright (c) 2017-2019 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package main
+
+import (
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"testing"
+
+	"github.com/decred/dcrtime/merkle"
+	"github.com/decred/politeia/politeiad/api/v1/identity"
+	"github.com/decred/politeia/politeiad/api/v1/mime"
+	cms "github.com/decred/politeia/politeiawww/api/cms/v1"
+	www "github.com/decred/politeia/politeiawww/api/www/v1"
+	"github.com/decred/politeia/politeiawww/user"
+	"github.com/decred/politeia/util"
+)
+
+func createValidLineItems(t *testing.T) []cms.LineItemsInput {
+	t.Helper()
+
+	// valid line item entries
+	lineItemLabor := cms.LineItemsInput{
+		Type:          cms.LineItemTypeLabor,
+		Domain:        "Development",
+		Subdomain:     "politeia",
+		Description:   "PR #1",
+		ProposalToken: "",
+		Labor:         40,
+		Expenses:      0,
+	}
+	lineItemExpense := cms.LineItemsInput{
+		Type:          cms.LineItemTypeExpense,
+		Domain:        "Design",
+		Subdomain:     "pgui",
+		Description:   "Artwork",
+		ProposalToken: "",
+		Labor:         0,
+		Expenses:      1000,
+	}
+	lineItemMisc := cms.LineItemsInput{
+		Type:          cms.LineItemTypeMisc,
+		Domain:        "Research",
+		Subdomain:     "dcrd",
+		Description:   "reorg",
+		ProposalToken: "",
+		Labor:         0,
+		Expenses:      10000,
+	}
+
+	return []cms.LineItemsInput{
+		lineItemLabor,
+		lineItemExpense,
+		lineItemMisc,
+	}
+}
+
+func createInvoiceInput(t *testing.T, li []cms.LineItemsInput) cms.InvoiceInput {
+	t.Helper()
+
+	var (
+		month       uint = 2
+		year        uint = 2019
+		rate        uint = 4000
+		monthAvg    uint = 1651
+		name             = "test"
+		location         = "testlocation"
+		contact          = "test@gmail.com"
+		paymentaddr      = "DsUHkmH555D4tLQi5ap4gVAV86tVN29nqYi"
+	)
+
+	return cms.InvoiceInput{
+		Version:            1,
+		Month:              month,
+		Year:               year,
+		ExchangeRate:       monthAvg,
+		ContractorName:     name,
+		ContractorLocation: location,
+		ContractorContact:  contact,
+		ContractorRate:     rate,
+		PaymentAddress:     paymentaddr,
+		LineItems:          li,
+	}
+}
+
+// createInvoiceJSON creates an index file with the passed InvoiceInput
+func createInvoiceJSON(t *testing.T, ii cms.InvoiceInput) *www.File {
+	t.Helper()
+
+	file, _ := json.Marshal(ii)
+
+	return &www.File{
+		Name:    invoiceFile,
+		MIME:    mime.DetectMimeType(file),
+		Digest:  hex.EncodeToString(util.Digest(file)),
+		Payload: base64.StdEncoding.EncodeToString(file),
+	}
+}
+
+// createNewInvoice computes the merkle root of the given files, signs the
+// merkle root with the given identity then returns a NewInvoice object.
+func createNewInvoice(t *testing.T, id *identity.FullIdentity,
+	files []www.File, month uint, year uint) *cms.NewInvoice {
```

Comment:
Don't split function declarations into multiple lines.  Function declarations can exceed the 80 column limit.

Reply:
Good to know that!